copier/copier.go

```diff
@@ -0,0 +1,1367 @@
+			}
+			hdr.Linkname = target
+			hdr.Size = 0
+		} else {
+			// note the device/inode pair for this file
+			hardlinkChecker.Add(srcfi, string(os.PathSeparator)+name)
+		}
+	}
+	// map the ownership for the archive
+	if idMappings != nil && !idMappings.Empty() {
+		hostPair := idtools.IDPair{UID: hdr.Uid, GID: hdr.Gid}
+		hdr.Uid, hdr.Gid, err = idMappings.ToContainer(hostPair)
+		if err != nil {
+			return errors.Wrapf(err, "error mapping host filesystem owners %#v to container filesystem owners", hostPair)
+		}
+	}
+	// output the header
+	if err = tw.WriteHeader(hdr); err != nil {
+		return errors.Wrapf(err, "error writing header for %s (%s)", contentPath, hdr.Name)
+	}
+	if hdr.Typeflag == tar.TypeReg {
+		// output the content
+		f, err := os.Open(contentPath)
+		if err != nil {
+			return errors.Wrapf(err, "error opening %s", contentPath)
+		}
+		defer f.Close()
+		n, err := io.Copy(tw, f)
+		if err != nil {
+			return errors.Wrapf(err, "error copying %s", contentPath)
+		}
+		if n != hdr.Size {
+			return errors.Errorf("error copying %s: incorrect size (expected %d bytes, read %d bytes)", contentPath, n, hdr.Size)
+		}
+		tw.Flush()
+	}
+	return nil
+}
+
+func copierHandlerPut(bulkReader io.Reader, request Request, idMappings *idtools.IDMappings) (*Response, func() error, error) {
+	errorResponse := func(fmtspec string, args ...interface{}) (*Response, func() error, error) {
+		return &Response{Put: PutResponse{Error: fmt.Sprintf(fmtspec, args...)}}, nil, nil
+	}
+	dirUID, dirGID := 0, 0
+	if request.PutOptions.ChownDirs != nil {
+		dirUID, dirGID = request.PutOptions.ChownDirs.UID, request.PutOptions.ChownDirs.GID
+	}
+	dirMode := os.FileMode(0755)
+	if request.PutOptions.ChmodDirs != nil {
+		dirMode = *request.PutOptions.ChmodDirs
+	}
+	var fileUID, fileGID *int
+	if request.PutOptions.ChownFiles != nil {
+		fileUID, fileGID = &request.PutOptions.ChownFiles.UID, &request.PutOptions.ChownFiles.GID
+	}
+	if idMappings != nil && !idMappings.Empty() {
+		containerDirPair := idtools.IDPair{UID: dirUID, GID: dirGID}
+		hostDirPair, err := idMappings.ToHost(containerDirPair)
+		if err != nil {
+			return errorResponse("copier: put: error mapping container filesystem owner %d:%d to host filesystem owners: %v", dirUID, dirGID, err)
+		}
+		dirUID, dirGID = hostDirPair.UID, hostDirPair.GID
+		if request.PutOptions.ChownFiles != nil {
+			containerFilePair := idtools.IDPair{UID: *fileUID, GID: *fileGID}
+			hostFilePair, err := idMappings.ToHost(containerFilePair)
+			if err != nil {
+				return errorResponse("copier: put: error mapping container filesystem owner %d:%d to host filesystem owners: %v", fileUID, fileGID, err)
+			}
+			fileUID, fileGID = &hostFilePair.UID, &hostFilePair.GID
+		}
+	}
+	ensureDirectoryUnderRoot := func(directory string) error {
+		rel, err := filepath.Rel(request.Root, directory)
+		if err != nil {
+			return errors.Wrapf(err, "%q is not a subdirectory of %q", directory, request.Root)
```

Comment:
That’s not what `filepath.Rel` does; this will happily write outside of `Root`.

Reply:
Aargh, missed that.  Prepending the path separator, running it through `filepath.Clean()` to remove initial "../" instances, and then trimming the prefixed path separator looks like it produces the desired result.